R/board_gcs.R

```diff
@@ -0,0 +1,262 @@
+  metadata_blob <- fs::path(name, version, "data.txt")
+
+  if (!gcs_file_exists(board, metadata_blob)) {
+    abort_pin_version_missing(version)
+  }
+
+  path_version <- fs::path(board$cache, name, version)
+  fs::dir_create(path_version)
+  gcs_download(board, metadata_blob)
+  local_meta(
+    read_meta(fs::path(board$cache, name, version)),
+    name = name,
+    dir = path_version,
+    version = version
+  )
+}
+
+#' @export
+pin_fetch.pins_board_gcs <- function(board, name, version = NULL, ...) {
+  withr::local_options(list(googleAuthR.verbose = 4))
+  meta <- pin_meta(board, name, version = version)
+  cache_touch(board, meta)
+
+  for (file in meta$file) {
+    key <- fs::path(name, meta$local$version, file)
+    gcs_download(board, key)
+  }
+
+  meta
+}
+
+#' @export
+pin_store.pins_board_gcs <- function(board, name, paths, metadata,
+                                     versioned = NULL, x = NULL, ...) {
+  withr::local_options(list(googleAuthR.verbose = 4))
+  ellipsis::check_dots_used()
+  check_name(name)
+  version <- version_setup(board, name, version_name(metadata), versioned = versioned)
+  version_dir <- fs::path(name, version)
+  gcs_upload_yaml(
+    board,
+    fs::path(paste0(board$prefix, version_dir), "data.txt"),
+    metadata
+  )
+
+  for (path in paths) {
+    googleCloudStorageR::gcs_upload(
+      file = path,
+      bucket = board$bucket,
+      name = fs::path(paste0(board$prefix, version_dir), fs::path_file(path)),
+      ...
+    )
+  }
+
+  name
+}
+
+#' @rdname board_deparse
+#' @export
+board_deparse.pins_board_gcs <- function(board, ...) {
+  bucket <- check_board_deparse(board, "bucket")
+  expr(board_gcs(!!bucket, prefix = !!board$prefix))
+}
+
+#' @rdname required_pkgs.pins_board
+#' @export
+required_pkgs.pins_board_gcs <- function(x, ...) {
+  ellipsis::check_dots_empty()
+  "googleCloudStorageR"
+}
+
+# Helpers -----------------------------------------------------------------
+
+gcs_delete_dir <- function(board, dir) {
+  resp <- googleCloudStorageR::gcs_list_objects(
+    bucket = board$bucket,
+    prefix = paste0(board$prefix, dir, "/")
+  )
+
+  if (nrow(resp) == 0) {
+    return(invisible())
+  }
+
+  for (path in resp$name) {
+    googleCloudStorageR::gcs_delete_object(path, bucket = board$bucket)
+  }
+
+  invisible()
+}
+
+gcs_upload_yaml <- function(board, key, yaml, ...) {
+  temp_file <- withr::local_tempfile()
+  yaml::write_yaml(yaml, file = temp_file)
+  googleCloudStorageR::gcs_upload(
+    file = temp_file,
+    bucket = board$bucket,
+    type = "text/yaml",
+    name = key,
+    ...
+  )
+}
+
+gcs_download <- function(board, key) {
+  path <- fs::path(board$cache, key)
+  if (!fs::file_exists(path)) {
+    suppressMessages(googleCloudStorageR::gcs_get_object(
+      object_name = paste0(board$prefix, key),
+      bucket = board$bucket,
+      saveToDisk = path
+    ))
+    fs::file_chmod(path, "u=r")
+  }
+  path
+}
+
+gcs_file_exists <- function(board, name) {
```

Comment:
I would do this via `gcs_get_object("obj_name", meta=TRUE)` and then `tryCatch(http_404 = FALSE)` - will be much quicker and direct only inspecting HTTP headers and not content.  Custom error types are supported in the HTTP responses to help this workflow. Listing objects could be slow for large amount of pins

Reply:
An example here for a pull request I did for targets a while ago https://github.com/MarkEdmondson1234/targets/blob/8204e4268553a2680f5d5ff3b0fc006b0c40d45a/R/utils_gcp.R#L7-22

Reply:
Ah, I see what you're saying! Hmmm... 

It's possible this function could have a better name, but it's not always looking for a specific object; sometimes it is looking for a directory-like thing. Given the information at hand when calling, for example, `pin_exists()`, I think we do need to list the objects in the bucket rather than get a specific object.